ludwig/config_validation/checks.py

```diff
@@ -111,6 +112,21 @@ def check_training_runway(config: "ModelConfig") -> None:  # noqa: F821
             )
 
 
+@register_config_check
+def check_optimizer_requirements(config: "ModelConfig") -> None:  # noqa: F821
+    """Checks if GPU is available when using Paged Optimizer."""
+    # TODO: Figure out why this check won't work
+    # Also it seems like this check is needed for 8-bit versions of the optimizers
+    # as well, so add a new property to the base class and then add that check here.
+    if not hasattr(config.trainer, "optimizer"):
+        return
+
+    if config.trainer.optimizer.is_paged and not torch.cuda.is_available():
```

Comment:
This isn't correct. GPU might be allocated later on (think Ray backend).

Reply:
Hmm, I'm trying to think of how we can catch this early enough and return a more straightforward error (the current one on CPU is quite bad and doesn't indicate the need to run this on GPUs)

Reply:
One alternative is to wrap optimizer initialization in a try-except, catch that specific error that's being thrown, and then raise a RuntimeError. 

But, from the code, it seems like we're not inside the Ray process yet at the time we make this call since we're still initializing the trainer, but we only have access to the right resources post training initialization once we make the call to trainer.train().

Reply:
This is the error that gets thrown btw:

```
NameError: name 'str2optimizer32bit' is not defined
```

Stack Trace:
```
def optimizer_update_32bit(
        optimizer_name: str,
        g: Tensor,
        p: Tensor,
        state1: Tensor,
        beta1: float,
        eps: float,
        step: int,
        lr: float,
        state2: Tensor = None,
        beta2: float = 0.0,
        weight_decay: float = 0.0,
        gnorm_scale: float = 1.0,
        unorm_vec: Tensor = None,
        max_unorm: float = 0.0,
        skip_zeros=False,
    ) -> None:
        """
        Performs an inplace optimizer update with one or two optimizer states.
    
        Universal optimizer update for 32-bit state and 32/16-bit gradients/weights.
    
        Parameters
        ----------
        optimizer_name : str
            The name of the optimizer: {adam}.
        g : torch.Tensor
            Gradient tensor.
        p : torch.Tensor
            Parameter tensor.
        state1 : torch.Tensor
            Optimizer state 1.
        beta1 : float
            Optimizer beta1.
        eps : float
            Optimizer epsilon.
        weight_decay : float
            Weight decay.
        step : int
            Current optimizer step.
        lr : float
            The learning rate.
        state2 : torch.Tensor
            Optimizer state 2.
        beta2 : float
            Optimizer beta2.
        gnorm_scale : float
            The factor to rescale the gradient to the max clip value.
        unorm_vec : torch.Tensor
            The tensor for the update norm.
        max_unorm : float
            The maximum update norm relative to the weight norm.
        skip_zeros : bool
            Whether to skip zero-valued gradients or not (default: False).
        """
    
        param_norm = 0.0
        if max_unorm > 0.0:
            param_norm = torch.norm(p.data.float())
    
    
        optim_func = None
        if g.dtype == torch.float32:
>           optim_func = str2optimizer32bit[optimizer_name][0]
E           NameError: name 'str2optimizer32bit' is not defined
```

Reply:
Okay I have an idea for a fix, let me push it up

Reply:
Optimizer is initialized here: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/distributed/base.py#L31

That's where you would make such a check.

Also, you should add a check for DeepSpeed here as well as it doesn't support these optimizers.

Reply:
Updated!